File: Core/src/main/java/com/github/ezauton/core/pathplanning/PP_PathGenerator.java
package com.github.ezauton.core.pathplanning;

import com.github.ezauton.core.pathplanning.purepursuit.PPWaypoint;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

/**
 * Generates a path 🗺 for Pure Pursuit given {@link PPWaypoint}
 */
public class PP_PathGenerator implements Serializable
{

    private final PPWaypoint[] ppWaypoints;

    public PP_PathGenerator(PPWaypoint... ppWaypoints)
    {
        this.ppWaypoints = ppWaypoints;
    }

    public Path generate(double dt)
    {
        List<IPathSegment> pathSegments = new ArrayList<>();
        double addedDistance = 0;
        for(int i = 0; i < ppWaypoints.length - 1; i++)
        {
            PPWaypoint from = ppWaypoints[i];
            PPWaypoint to = ppWaypoints[i + 1];

            //TODO: Update from RobotCode2018 style pathsegments
            LinearPathSegment pathSegment;
            if(i == 0)
            {
                double beginningSpeed = from.getSpeed() == 0 ? to.getSpeed() : from.getSpeed();


                pathSegment = new PathSegmentInterpolated(
                        from.getLocation(), to.getLocation(), i == ppWaypoints.length - 2, true, addedDistance,
                        beginningSpeed, to.getSpeed(), dt,
                        from.getAcceleration(), from.getDeceleration());
            }
            else
            {
                pathSegment = new PathSegmentInterpolated(
                        from.getLocation(), to.getLocation(), i == ppWaypoints.length - 2, false, addedDistance,
                        from.getSpeed(), to.getSpeed(), dt,
                        from.getAcceleration(), from.getDeceleration());
            }
            addedDistance += pathSegment.getLength();
            pathSegments.add(pathSegment);
        }
        return Path.fromSegments(pathSegments);
    }
}

File: Core/src/main/kotlin/com/github/ezauton/core/pathplanning/PP_PathGenerator.kt
package com.github.ezauton.core.pathplanning

//import com.github.ezauton.core.pathplanning.purepursuit.PPWaypoint
import java.io.Serializable

///**
// * Generates a path 🗺 for Pure Pursuit given [PPWaypoint]
// */
//class PP_PathGenerator(vararg ppWaypoints: PPWaypoint) : Serializable {
//
//  private val ppWaypoints: Array<PPWaypoint>
//
//  init {
//    this.ppWaypoints = ppWaypoints
//  }
//
//  fun generate(dt: Double): Path {
//    val pathSegments = ArrayList<PathSegment>()
//    var addedDistance = 0.0
//    for (i in 0 until ppWaypoints.size - 1) {
//      val from = ppWaypoints[i]
//      val to = ppWaypoints[i + 1]
//
//      // TODO: Update from RobotCode2018 style pathsegments
//      val pathSegment: LinearPathSegment
//      pathSegment = if (i == 0) {
//        val beginningSpeed = if (from.speed == 0.0) to.speed else from.speed
//
//        PathSegmentInterpolated(
//          from.location, to.location, i == ppWaypoints.size - 2, true, addedDistance,
//          beginningSpeed, to.speed, dt,
//          from.acceleration, from.deceleration
//        )
//      } else {
//        PathSegmentInterpolated(
//          from.location, to.location, i == ppWaypoints.size - 2, false, addedDistance,
//          from.speed, to.speed, dt,
//          from.acceleration, from.deceleration
//        )
//      }
//      addedDistance += pathSegment.length
//      pathSegments.add(pathSegment)
//    }
//    return Path.fromSegments(pathSegments)
//  }
//}
